Added toolchain abstraction for coverage

vanwinkeljan opened the pull request:
~~Added compiler and linker flags to generate gcov compatible coverage results with clang.~~

Added toolchain abstraction for coverage for both gcc and clang.

Note that without this support sanitycheck can not run without error when llvm toolchain is used.

vanwinkeljan committed:
cmake: toolchain abstraction for coverage

Added toolchain abstraction for coverage for both gcc and clang.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

aescolar reviewed (dismissed):
I'd say we should be moving this into 
cmake/compiler/host-gcc/target.cmake
cmake/compiler/clang/target.cmake
under and `if ("${ARCH}" STREQUAL "posix")`

SebastianBoe commented on arch/posix/CMakeLists.txt:
```diff
@@ -30,6 +32,17 @@ zephyr_link_libraries_ifdef(CONFIG_COVERAGE
+	--coverage
+)
+endif()
+
```
We use a different mechanism to achieve portability.

See

https://github.com/zephyrproject-rtos/zephyr/commit/01592071f1e205e7a020d9dd487c04943226c287

for an example.

SebastianBoe commented:
> under and if ("${ARCH}" STREQUAL "posix")

If the coverage API is invoked from arch/posix/CMakeLists.txt (which it should be), then an arch check is not necessary.

aescolar commented:
> If the coverage API is invoked from arch/posix/CMakeLists.txt (which it should be), then an arch check is not necessary.

True story :), I guess we just need to be careful to not mix it with the coverage for target

SebastianBoe commented:
The coverage for target?

aescolar commented:
> The coverage for target?

Support for `CONFIG_COVERAGE` with "real targets" was added a couple of months ago:
https://docs.zephyrproject.org/latest/guides/coverage.html#test-coverage-reports-in-embedded-devices-or-qemu

vanwinkeljan commented:
@SebastianBoe @aescolar  Added toolchain abstraction for coverage for both host and target

vanwinkeljan committed:
sanitycheck: native: Removed --coverage from EXTRA_LDFLAGS

Removed redundant linker flag --coverage from native posix
EXTRA_LDFLAGS.

The correct linker flags will be set by defining the Kconfig flag
CONFIG_COVERAGE, either by sanitcheck it self or via the test/sample
prj.conf.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan commented:
Updated PR to remove redunat linker option set in sanitycheck script

aescolar reviewed (commented):
No complaints from me
The sanitycheck change is perfectly ok.
But as this is mostly cmake related, I let @SebastianBoe have the final say

SebastianBoe commented on cmake/compiler/clang/target_coverage.cmake:
```diff
@@ -0,0 +1,22 @@
+
+if (CONFIG_COVERAGE)
+
+zephyr_compile_options(
```
2 spaces of indentation.

SebastianBoe commented on cmake/compiler/clang/target_coverage.cmake:
```diff
@@ -0,0 +1,22 @@
+
+macro(toolchain_cc_coverage)
+
+if (CONFIG_COVERAGE)
```
This if-statement belongs in the CMakeLists.txt files.

SebastianBoe reviewed (changes requested):
Requesting changes

vanwinkeljan commented:
open/closing to kick CI
